### PR TITLE
send queue: allow editing local echoes

### LINKED ISSUE #3361: timeline: introduce a Send Queue API
- [x] Make it so that the sending queue is shared by all the timeline instances of the same room (live or detached), so that local echoes appear on all the live/detached timelines, if need be
- [x] Allow retry sending
- [x] Allow enabling/disabling the sending queue globally
- [x] Change course to have a per-room sending queue enablement flag
- [x] Expose whenever an event couldn't be sent because of an API error (status code 4XX or 5XX)
- [x] Mark such events from the queue as wedged, and don't block the queue for them.
- [x] Store and manage unsent events on disk
  - [x] make all the ruma `*EventContent` types serializable and deserializable with a type ([PR](https://github.com/ruma/ruma/pull/1850))
  - [x] [implement state store backends to save events on disk](https://github.com/matrix-org/matrix-rust-sdk/pull/3592)
  - [x] reload all the rooms that have at least one pending event when the Client restarts
  - [x] timeline: mark wedged local echoes with the proper failure send state
- [x] Allow redacting local echoes in a timeline
- [x] Allow editing local echoes in a timeline
- [x] Allow sending raw events
  - blocked by https://github.com/ruma/ruma/issues/1865

---

* Edit by @Hywan: Fix https://github.com/element-hq/element-x-ios/issues/1758

### PR DESCRIPTION
**Pull request by bnjbvr**
Final step of #3361, as a bonus.

Fixes #3361.

### CODE REVIEW

File: crates/matrix-sdk/src/send_queue.rs [hunk 1]
```diff
@@ -698,6 +698,32 @@ impl QueueStorage {
...
+        if self.being_sent.read().await.contains(transaction_id) {
+            return Ok(false);
+        }
+
+        self.client
+            .get()
+            .ok_or(RoomSendQueueStorageError::ClientShuttingDown)?
+            .store()
+            .update_send_queue_event(&self.room_id, transaction_id, serializable)
+            .await?;
```
Comment (Hywan) [inline crates/matrix-sdk/src/send_queue.rs:722]: Can we have a race condition here? I think no, but maybe `being_sent` can be modified before `update_send_queue_event` is called?

File: crates/matrix-sdk-ui/src/timeline/inner/mod.rs [hunk 2]
```diff
@@ -805,6 +805,53 @@ impl<P: RoomDataProvider> TimelineInner<P> {
...
 
+    pub(super) async fn replace_local_echo(
+        &self,
+        txn_id: &TransactionId,
+        content: AnyMessageLikeEventContent,
+    ) -> bool {
+        let AnyMessageLikeEventContent::RoomMessage(content) = content else {
+            // Ideally, we'd support replacing local echoes for a reaction, etc., but
+            // handling RoomMessage should be sufficient in most cases. Worst
+            // case, the local echo will be sent Soon™ and we'll get another chance at
```
Comment (Hywan) [inline crates/matrix-sdk-ui/src/timeline/inner/mod.rs:816]: “Soon™” 😄  

By the way, I agree with this. It means someone reacts to its own local event, and wants to change that. That's clearly an edge case.

File: bindings/matrix-sdk-ffi/src/timeline/mod.rs [hunk 3]
```diff
@@ -470,10 +470,36 @@ impl Timeline {
...
+    /// redacted by sending an edit request to the server.
+    ///
+    /// Returns whether the edit did happen. It can only return false for
+    /// local events that are being processed.
     pub async fn edit(
         &self,
+        item: Arc<EventTimelineItem>,
         new_content: Arc<RoomMessageEventContentWithoutRelation>,
+    ) -> Result<bool, ClientError> {
+        let edit_info = item.0.edit_info().map_err(|err| anyhow::anyhow!(err))?;
```
Comment (Hywan) [inline bindings/matrix-sdk-ffi/src/timeline/mod.rs:488]: `ClientError::from(err)` doesn't work (to replace `anyhow!`)?

**Review by Hywan — APPROVED** [review]
It looks generally good to me. The fact the patches modify both the `SendQueue`, the database, and the `Timeline` make it a bit difficult to follow, but thanks to your atomic commits, it wasn't that much complicated.

Please rebase your `fixup!` patches before merging your PR.

File: crates/matrix-sdk/src/send_queue.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/send_queue.rs:722]: No indeed: there's only one possible caller, the sender task. Will add a small comment.

File: bindings/matrix-sdk-ffi/src/timeline/mod.rs (hunk 3 above)
Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/timeline/mod.rs:488]: Yes, with a new conversion method. I'll add it. We should do a pass at making better errors at the FFI layer some day.

**Pull request merged**